source/agora/network/NetworkManager.d

```diff
@@ -57,31 +57,290 @@ mixin AddLogger!();
+                                     node and the node was banned
+
+        ***********************************************************************/
+
+        public this (in Address address,
+            void delegate (scope ref NodeConnInfo node) onHandshakeComplete,
+            void delegate (in Address address) onFailedConnection)
+        {
+            this.address = address;
+            this.onHandshakeComplete = onHandshakeComplete;
+            this.onFailedConnection = onFailedConnection;
+            this.outer.taskman.runTask(&this.connect);
+        }
+
+        /***********************************************************************
+
+            Repeatedly attempt connecting to an address, and try to determine
+            if this is a FullNode or a Validator.
+
+            After each connection failure / request failure the ban manager is
+            queried, and if this address was banned then 'onFailedConnection()'
+            will be called and the task will be killed.
+
+            If we've successfully connected and determined the Node type,
+            'onHandshakeComplete' is called and the task is killed.
+
+        ***********************************************************************/
+
+        private void connect ()
+        {
+            API api;
+            while (api is null)
+            {
+                try
+                {
+                    api = this.outer.getClient(this.address,
+                        this.outer.node_config.timeout.msecs);
+                }
+                catch (Exception ex)
+                {
+                    if (this.outer.banman.isBanned(this.address))
+                    {
+                        this.onFailedConnection(this.address);
+                        return;
+                    }
+
+                    this.outer.taskman.wait(
+                        this.outer.node_config.retry_delay.msecs);
+                }
+            }
+
+            auto client = new NetworkClient(this.outer.taskman,
+                this.outer.banman, this.address, api,
+                this.outer.node_config.retry_delay.msecs,
+                this.outer.node_config.max_retries);
+
+            PublicKey key;
+            while (1)
+            {
+                try
+                {
+                    // LocalRest will return PublicKey.init,
+                    // vibe.d will throw HTTPStatusException with status == 404
+                    key = client.getPublicKey();
+                    break;
+                }
+                catch (HTTPStatusException ex)
+                {
+                    // 404 => API not implemented, it's a FullNode
+                    if (ex.status == 404)
+                        break;
+
+                    if (this.outer.banman.isBanned(this.address))
+                    {
+                        this.onFailedConnection(this.address);
+                        return;
+                    }
```

Comment:
My inner developer suggests me that we might want to make `onFailedConnection` returns a boolean telling whether we should retry or abort the request, and that we *could* move the `isBanned` check in there.
I don't know how sensible this would be, and I'm not asking you to do it, just raising possibilities.

Reply:
Oh this is a good idea.